Allow indicators to take up top panel (and also fix indicators)

UniversalSuperBox opened the pull request:
The addition of application menus made it more difficult to hit the target area for the messages indicator. Prior to the application menu addition, the entire blank area of the panel could be used. After, only the 2 gu space taken up by the icon could be used.

This was due to an implementation detail of the application menus... when the Shell has 60 gu or less width, both the application menus and the indicators are set to the full width of the panel. This meant that their drag handles, the area where they could be "pulled down" from, took the entire panel. Since the indicators were defined after the application menus in the file, the indicators' handle went over the application menus' and the menus were unusable.

To solve this, the original author forced the indicators' handle to only use the space taken by the indicator items. This approach had two big problems. First, if the user had enough indicators to reach all the way across the screen, the application menus would be unusable. Second, it shrank the usable handle area for the leftmost indicator.

This change adjusts the indicator bar to provide a 7 grid unit space to its left whenever the application menus are available. In case of space contention between the two (read: too many indicators are showing), the leftmost indicators will hide. Whenever the application menus are not available, the indicators (and, more importantly, their handle) take up the full width of the screen again.

Additionally, I added a "down" icon to the panel. It shows up at any screen size, but only when application menus are available. This should help improve discoverability of the application menus, which were visually "silent" in Staged mode prior to this change.

![Indicators and panel sharing phone-sized area](https://user-images.githubusercontent.com/21966173/75909215-413a0d80-5e11-11ea-8c40-b9a5dcab5a35.png)

![Indicators and panel sharing very small area, showing indicators overflowing during space contention](https://user-images.githubusercontent.com/21966173/75909233-4ac37580-5e11-11ea-8655-e593a43d2490.png)

![Indicators and panel sharing a large area, showing window title appearing with "down" icon](https://user-images.githubusercontent.com/21966173/75909289-7181ac00-5e11-11ea-9562-33bbdb155f9d.png)


Thanks @peat-psuwit for figuring out my mistakes and repainting over them to actually finish this PR.

Please rebase and merge or use a merge commit, the commit history is useful.

Fixes ubports/ubuntu-touch#1233
Fixes ubports/ubuntu-touch#1364

UniversalSuperBox committed:
Allow indicators to take up more space

The addition of application menus made it more difficult to hit the
target area for the messages indicator. Prior to the application menu
addition, the entire blank area of the panel could be used. After, only
the 2 gu space taken up by the icon could be used.

This was due to an implementation detail of the application menus...
when the Shell has 60 gu or less width, both the application menus and
the indicators are set to the full width of the panel. This meant that
their drag handles, the area where they could be "pulled down" from,
took the entire panel. Since the indicators were defined after the
application menus in the file, the indicators' handle went over the
application menus' and the menus were unusable.

To solve this, the original author forced the indicators' handle to only
use the space taken by the indicator items. This approach had two big
problems. First, if the user had enough indicators to reach all the way
across the screen, the application menus would be unusable. Second, it
shrank the usable handle area for the leftmost indicator.

This change adjusts the indicator bar to provide a 7 grid unit space
to its left whenever the application menus are available. In case of
space contention between the two (read: too many indicators are
showing), the leftmost indicators will hide. Whenever the application
menus are not available, the indicators (and, more importantly, their
handle) take up the full width of the screen again.

Additionally, I added a "down" icon to the panel. It shows up at any
screen size, but only when application menus are available. This should
help improve discoverability of the application menus, which were
visually "silent" in Staged mode prior to this change.

Finally, I've noticed that sometimes tests may not clean up properly
after setting indicators to hidden. To fix this, I manually set the
indicators hidden then visible again. Since I needed a test case for
a smaller screen and didn't want to copy so much code, I refactored
the test case into a new PanelUI file for the setup.

Fixes https://github.com/ubports/ubuntu-touch/issues/1233

UniversalSuperBox committed:
Add button to show or hide app menus to PanelTests

peat-psuwit committed:
Compensate for PanelItemRow's origin changing

The QtQuick Flickable documentation [1] states that, "If you flick back
to the origin, after the rebound animation, contentX will settle to the
same value as originX. These are usually (0,0), however ListView and
GridView may have an arbitrary origin..."

This was causing problems for us in the Indicators. Some events, such as
a manual 'show()' of the PanelMenu, could cause the ListView's origin to
shift. When this occured, the highlight would be incorrect and other
interactions would not work correctly. For example, trying to select the
rightmost indicator would cause the leftmost to be selected.

To fix this, use the PanelItemRow's contentX to compensate some values
of X for the difference.

Fixes https://github.com/ubports/ubuntu-touch/issues/1364

[1] https://doc.qt.io/qt-5/qml-qtquick-flickable.html#contentX-prop

peat-psuwit committed:
Fix Panel tests freezing occasionally

During Panel tests, it was common to see the following error repeating
in the log:

QWARN  : [...]unity8/qml/Panel/PanelItemRow.qml:168:5: QML ListView: Binding loop detected for property "width"

Generally this didn't cause any problems. Occasionally, though, the loop
would not be caught and the tests would lock up forever at 100% CPU.

To fix this, use the ListView's contentItem rather than childrenRect.
This likely should have been the case since the beginning, since the
ListView reparents all contained Items to contentItem rather than
holding them directly.

mardy commented on qml/Panel/Panel.qml:
```diff
@@ -325,26 +328,53 @@ Item {
+                width: root.width - __indicators.maxDragHandleSize - touchMenuIcon.width
+                height: root.minimizedPanelHeight
+                verticalAlignment: Text.AlignVCenter
+                elide: Text.ElideRight
```
Can the text be truncated? This property seems to imply that it can, but some other parts of the code make me think that the item is always being allocated the `implicitWidth` it requested.

UniversalSuperBox replied:
![image](https://user-images.githubusercontent.com/21966173/76371883-bcb42700-6309-11ea-856d-634dd0102eca.png)

oh.

UniversalSuperBox replied:
![image](https://user-images.githubusercontent.com/21966173/76375071-e756ad80-6312-11ea-98f3-667f0aa27209.png)

This is fixed with the latest commit, as seen here. The gray rectangles are showing the handles which can be used to drag the menus down.

mardy commented on qml/Panel/Panel.qml:
```diff
@@ -325,26 +328,53 @@ Item {
+                objectName: "touchMenuIcon"
+                anchors {
+                    right: parent.right
+                    rightMargin: parent.width - rowLabel.implicitWidthWithMargin - width
```
This code assumes that the `rowLabel` is always allocated the full `implicitWidthWithMargin`, but the `rowLabel`'s `width` property seems to be computed in some other way.

UniversalSuperBox replied:
I believe this is fixed with the newest commit

mardy commented on qml/Panel/Panel.qml:
```diff
@@ -325,26 +328,53 @@ Item {
+                id: touchMenuIcon
+                objectName: "touchMenuIcon"
+                anchors {
+                    right: parent.right
```
why isn't this anchored to the `rowLabel`'s right? Then the `rightMargin` property could be left unspecified.

UniversalSuperBox replied:
Did exactly that

mardy commented on qml/Panel/Panel.qml:
```diff
@@ -37,8 +38,7 @@ Item {
     property real expandedPanelHeight: units.gu(7)
-    property real indicatorMenuWidth: width
-    property real applicationMenuWidth: width
+    property real menuWidth: partialWidth ? units.gu(40) : width
```
Am I understanding correctly, that with the old code, each of the menus could fill up the whole panel, whereas now, if the screen is large enough, each of them would be limited to 40 gu? So, if my screen would be 120 gus, only 40 gus on each side would be used, whereas the central 40 will always be empty?

UniversalSuperBox replied:
In the old code, there were no application menus. Therefore there was no need to make the indicators get out of the way of the application menus. Your assumption is correct, at least for the touch menus. No matter how wide your screen is, only 40gu is allocated for each menu.

![image](https://user-images.githubusercontent.com/21966173/76375325-87acd200-6313-11ea-9743-ae88a1904b8a.png)

(Shown with rectangles over the touch areas for clarity)

It's done this way because someone decided that the menu width should be the same as the handle width.

![image](https://user-images.githubusercontent.com/21966173/76375357-985d4800-6313-11ea-9e78-c93e3a477b36.png)

Global menus for mouse can take up more of the panel, and fade in when the mouse is over the panel. Similar to Unity7's approach.

I agree this is an issue, but since no known Ubuntu Touch apps use the application menus it should not be a priority right now.

mardy reviewed (changes requested):
I left a few comments, but mostly to understand how things work. Honestly, to me this code is mostly unreadable, and it looks terribly hard to maintain.

It may be that some of the things I'm suggesting are already implemented and I just failed to recognize them, but the way I'd proceed would be this:

1. Create a new item which has two children: the `__applicationMenus` panel and the `panelTitleHolder`. It will have an `implicitWidth` property set to the max of the desired widths of these two children, and when it gets its width set, it will enforce it on the children to make sure they won't escape it.

2. Add an `implicitWidth` property to the `__indicators` panel.

3. Make sure that the two items above (the new item and the `__indicators` one) expose all the properties that are needed to make a decision about their size and visibility.

4. Create a new `QtObject` that will take the two items above as properties and has the logic to compute their final width, and sets it on the items. Ideally (especially if this code is non trivial), move this QtObject to its own QML file, so that it will be much simpler to write unit tests for it.

5. In general, don't be shy in moving complex items into their own QML file.

As for the user experience itself, when there is not enough width to show the menus, I'd also shrink the indicator, but add a "..." symbol to make it clear that we are hiding some of them. And when the user starts a drag operation over the indicators area, they should get expanded and become all visible (maybe this is already the case).

UniversalSuperBox committed:
Fix eliding of title

Previous changes prevented the panelTitle's text from eliding to get out
of the way of the indicators. By setting the anchors properly, we can
fix that issue.

peat-psuwit commented on qml/Panel/Panel.qml:
```diff
@@ -48,6 +48,9 @@ Item {
     property bool hasKeyboard: false
 
+    // Whether our expanded menus should take up the full width of the panel
+    property bool partialWidth: width > units.gu(60)
```
This will become false when `width` is 1 pixel above 60 GUs. Is this intentional? Or, what is the correct approach here? Does it matter?

UniversalSuperBox replied:
There's really no other way to make breakpoints. There will always be a one-pixel spot where you can cause the layout to change.

peat-psuwit replied:
I mean, is this supposed to be `>=` instead of `>`? In other words, which one-pixel spot should we consider?

UniversalSuperBox replied:
I'll use `>=`.

peat-psuwit reviewed (commented):
This is pretty large. So, here's my understanding:
- `PanelMenu`'s `overflowWidth` dictates how wide is the handle. However, this value can't be wider than the `width` itself.
- If handles overlap, whatever is on the top (declared later) get the touch.
- When the screen is small, both app menu and indicator bar take the whole screen width. However, indicator bar will retract `overflowWidth` by 7 GUs when app menu is enabled, enabling app menu to be interacted.
- However, when the screen is large (>60 GUs), each of them will take only 40 GUs. In this case, it seems like the 7 GU clearance is not given as it's not needed. The handles might still overlap, but the app menu is guaranteed to have at least 20 GUs to interact with.
- The window title is a separated entity, visible when the screen is large, and expand until the starts of indicator panel.

I have a pedantic comment below. But otherwise, LGTM but someone else must approve.

UniversalSuperBox committed:
Make partial panel width exclusive to 60 gu

Flohack74 commented:
@mardy can you do a quick summary review here, or should someone else take this over?

mardy commented:
> @mardy can you do a quick summary review here, or should someone else take this over?

Well, as I wrote [above](https://github.com/ubports/unity8/pull/289#pullrequestreview-372226669), there are several changes I'd suggest to make, but I don't think it should prevent this MR from being merged. Code-wise, it's not easy to understand, but on the other hand this is mostly due to the existing code -- this branch does only add a little drop to the ocean. :-)

I'd invite people from the community to test the PPA where this branch is built to and, if no regressions are found, just merge it.

UniversalSuperBox commented:
The changes to tests are much larger than the changes to production code, so this isn't quite as scary as " +522 −262" would have you believe.

Balancing mobile and desktop usability with the current reality of releasing OTA-12 and the requested refactoring's difficulty, I believe this should be merged.

To be more detailed about a refactoring, Mardy's requested changes aren't all that have been requested. @peat-psuwit also commented that a ListView inside a Flickable is a poor choice here, and leads to a lot of problems.

peat-psuwit reviewed (approved):
The new commit invalidates my review. So, again, LGTM, but someone else must approve. If @mardy agrees with me, please add another approval.
